Handle multiple servers per language

gbrlsnchs opened the pull request:
This PR is my first formal attempt of pushing necessary changes for handling multiple language servers per filetype.

The code is still broken and doesn't compile. I'll keep pushing commits here when I have time to work on this, preferably organized in meaningful commits, in order to track progress and promote discussion over how it is looking.

Also, I'll keep updating this description as new questions arise. Hopefully this PR will serve as a kick-start for solving this long-standing issue (only if desired).

Fixes #17.

To do:
- [x] Make it work properly
- [x] Tidy commits
- [x] Remove labels for single server
- [ ] Sort responses by priority
- [x] Update config file
- [x] Update docs
- [x] Make sure goto buffers are properly set when 2+ servers have different root paths

---

### Configuring multiple language servers

```toml
[language_ids]
javascript = "javascriptreact"
typescript = "typescriptreact"

[language_server.tsserver]
filetypes = ["javascript", "typescript"]
roots = ["package.json", "tsconfig.json", "jsconfig.json", ".git", ".hg"]
command = "typescript-language-server"
args = ["--stdio"]

[language_server.tailwindcss]
filetypes = ["javascript", "typescript"]
roots = ["tailwind.config.js", "tailwind.config.ts"]
command = "tailwindcss-language-server"
args = ["--stdio"]

[language_server.tailwindcss.settings.tailwindcss]
editor = {}
```

> **Note**
> You can now either use `language` or `language_server` as the field name. However, it should be consistent across your config.

## Screenshots

### Capabilities

![image](https://github.com/kak-lsp/kak-lsp/assets/7577128/30493185-627f-4a65-bb95-d1c4bc9b4ec0)

### Formatting (and range formatting)

When there are more than one capable formatting server, a prompt appears so users can choose which formatting server to use (same prompt used in code actions/lenses).

![image](https://github.com/kak-lsp/kak-lsp/assets/7577128/6237c62e-a676-4b9c-83a5-f728b6de91e4)

sclu1034 commented:
A heads-up, as you might want to keep this in consideration early, depending on how apt you're with amending and rebasing:

In the interest of keeping `git blame` useful and sane, every individual commit should successfully compile by itself.
It's not as important while the PR is still in draft, but the above should be ensured before it is merged.

krobelus commented on src/context.rs:
```diff
@@ -33,76 +34,91 @@ pub struct OutstandingRequests {
+    pub code_lenses: HashMap<String, Vec<(ServerName, CodeLens)>>,
+    pub completion_items: Vec<(ServerName, CompletionItem)>,
     pub completion_items_timestamp: i32,
     // We currently only track one client's completion items, to simplify cleanup (else we
```
Impressive work! A couple of comments (I didn't review the whole thing yet):

To optimzie the changes for the reader, can you structure the history so that every commit passes all tests (while being as small as practically possible).
See https://www.bitsnbites.eu/git-history-work-log-vs-recipe/
In particular, I don't care about any "fix" commits that fix bugs in your earlier commits. They should be squashed.

I'd probaly make one commit that extracts the `ServerSettings` type (and makes sure everything still compiles) without changing any behavior.
Then the actual logic changes will be much smaller.

> [language.typescriptreact]

While this may work in practice, I don't think this is kosher.
The "typescriptreact" part is sent to the language server as language ID.
The default config use "jsx" and "tsx" because those happen to work for "javascript" resp. "typescript" with the language servers I tried.

> Breaking changes

Don't. An old config file should keep working. Perhaps we want to make the parser accept both styles.

> [language.typescriptreact.tsserver]
> filetypes = ["javascript", "typescript"]
> roots = ["package.json", "tsconfig.json", "jsconfig.json", ".git", ".hg"]
> command = "typescript-language-server"
> args = ["--stdio"]
> 
> [language.typescriptreact.tailwindcss]
> filetypes = ["javascript", "typescript"]
> roots = ["tailwind.config.js", "tailwind.config.ts"]
> command = "tailwindcss-language-server"
> args = ["--stdio"]
> settings_section = "tailwindCSS"

what's wrong with

	[language.tsserver]
	filetypes = ["javascript", "typescript"]
	...
	[language.tailwindcss]
	filetypes = ["javascript", "typescript"]
	...

gbrlsnchs replied:
> Impressive work! A couple of comments (I didn't review the whole thing yet):
> 
> To optimzie the changes for the reader, can you structure the history so that every commit passes all tests (while being as small as practically possible). See https://www.bitsnbites.eu/git-history-work-log-vs-recipe/ In particular, I don't care about any "fix" commits that fix bugs in your earlier commits. They should be squashed.

Sure thing! I just wanted to wait for feedback on some decisions, so I can work around tests based on that feedback. Once I get tests passing I'll squash everything together.

> I'd probaly make one commit that extracts the `ServerSettings` type (and makes sure everything still compiles) without changing any behavior. Then the actual logic changes will be much smaller.

This sounds good, will try it!

> > [language.typescriptreact]
> 
> While this may work in practice, I don't think this is kosher. The "typescriptreact" part is sent to the language server as language ID. The default config use "jsx" and "tsx" because those happen to work for "javascript" resp. "typescript" with the language servers I tried.

What led me to use that was the TailwindCSS server not working without a proper language ID. Then I followed the [suggested values in the specification]( https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentItem) and it started working.

> > Breaking changes
> 
> Don't. An old config file should keep working. Perhaps we want to make the parser accept both styles. (EDIT: FWIW an example of a deserializer that accepts two flavors of TOML structures is [0c10b3c](https://github.com/kak-lsp/kak-lsp/commit/0c10b3c9d072c8e3af0e217f5ee54e31a231dc62). Thankfully we never needed to merge that. Maybe we also don't need ambiguity here)
> 
> > [language.typescriptreact.tsserver]
> > filetypes = ["javascript", "typescript"]
> > roots = ["package.json", "tsconfig.json", "jsconfig.json", ".git", ".hg"]
> > command = "typescript-language-server"
> > args = ["--stdio"]
> > [language.typescriptreact.tailwindcss]
> > filetypes = ["javascript", "typescript"]
> > roots = ["tailwind.config.js", "tailwind.config.ts"]
> > command = "tailwindcss-language-server"
> > args = ["--stdio"]
> > settings_section = "tailwindCSS"
> 
> what's wrong with
> 
> ```
> [language.tsserver]
> filetypes = ["javascript", "typescript"]
> ...
> [language.tailwindcss]
> filetypes = ["javascript", "typescript"]
> ...
> ```

The problem is that we need to use a "valid" language ID, because some servers won't work without them (as mentioned, TailwindCSS only works with specific language IDs from [the specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentItem)).

I agree breaking a user's config is bad, and I would prefer an approach that doesn't as well. I thought we could go with something like:
```toml
[language.tsserver]
language_id = "typescriptreact"
filetypes = ["javascript", "typescript"]

[language.tailwindcss]
language_id = "typescriptreact"
filetypes = ["javascript", "typescript"]
```

In the example above, `language_id` would be optional, and if not present, the `language.<name>` would be used instead (like it currently is).

My biggest gripe with that approach is terminology, as `language` was previously attached to a language ID, and now it is attached to a language server name.

What I propose is that we have two hashmaps --- `language` and `language_server` --- and that they get merged together when config loads. This way we can avoid a breaking change by making `language` and `language_server` aliases, while also allowing the correct terminology to be used in newer versions of the config.

Here's an example of a valid config:

```toml
# This is the old config the user already has,
# and this shall not break by any means.
[language.zig]
filetypes = ["zig"]
command = "zls"

# This is the new config the user can now use.
[language_server.tsserver]
language_id = "typescriptreact"
filetypes = ["javascript", "typescript"]
command = "..."
```

krobelus replied:
your proposal with `language_server` instead of `language` makes a ton of sense - there can be multiple servers for the same language.

krobelus replied:
you're right, `typescriptreact` is the correct language ID. Here's an idea, maybe we want to write

```toml
[[language.typescriptreact]]
command = "typescript-language-server"
...
[[language.typescriptreact]]
command = "tailwindcss-language-server"
...
...
```

haven't tried it so not sure

gbrlsnchs replied:
> you're right, typescriptreact is the correct language ID. Here's an idea, maybe we want to write
>
> ```toml
> [[language.typescriptreact]]
> command = "typescript-language-server"
> ...
> [[language.typescriptreact]]
> command = "tailwindcss-language-server"
> ...
> ...
> ```
> haven't tried it so not sure

This one is elegant. The only change we'd need to make is adding an optional `server` field, so we can identify each server. It needs to be optional in order to still be backwards compatible. If no server name is informed, we can use the language ID, and if it's a list, we can append an index to it (resulting in `(typescriptreact-1, typescriptreact-2)` for example). It's ugly but not the end of the world, and we can always provide examples in docs explaining that passing a `server` makes info output better. Such behavior can be achieved by using a custom deserializer, allowing the following:

```toml
# Existing config.
[language.rust]
filetypes = ["rust"]
command = "rust-analyzer"

[[language.typescriptreact]]
server = "tsserver"
filetypes = ["javascript", "typescript"]
command = "typescript-language-server"
args = ["--stdio"]

[[language.typescriptreact]]
server = "tailwindcss"
filetypes = ["javascript", "typescript"]
command = "tailwindcss-language-server"
args = ["--stdio"]
```

gbrlsnchs replied:
@krobelus I tried implementing the approach above and it doesn't work well with TOML, specially when defining `settings` for a language server.

I think maybe the aliasing approach is simpler and easier to maintain.

gbrlsnchs replied:
@krobelus I tried the approach above and 1) it didn't work properly and 2) is difficult to implement and hard to maintain.

Maybe we could go with the aliasing approach? It's still backwards compatible and is a lot cleaner.

krobelus replied:
I don't see why it wouldn't work but we can handle this later. It doesn't block other work and changes to this will be small and isolated

krobelus commented on src/language_features/formatting.rs:
```diff
@@ -1,31 +1,77 @@
+                format!("{} {}", editor_quote(server_name), editor_quote(&cmd))
+            })
+            .join(" ");
+        ctx.exec(meta, format!("lsp-menu {}", choices));
```
I'm not sure about this. Probably the user should decide this earlier.
Also we'll want a non-interactive version for hooks.

The same problem exists for many other requests besides formatting, right?

Need to think about which ones have such conflicts.
An alternative solution is to have a primary server for such ambiguous requests.
I wonder what fellow LSP clients do.

gbrlsnchs replied:
Indeed, although rarer for other methods and much more common for formatting, this race condition problem exists.

Note that, if only a single server is used, then there is no prompt anyway, so this would be a breaking change for users.

IMO, having a prompt for picking the formatting server is a good default UX for a multi-server setup (just reminding that the prompt won't be shown to most users, who will still use a single server).

Having priorities configured is also good UX IMO, and I think both can co-exist if the user doesn't want a prompt or if the user doesn't want to set up `lsp-formatting <server>` for languages that use multiple servers.

For users that desire to use `lsp-formatting` and the likes in a non-interactive way with multiple servers, one can inform such commands the desired server (for example `lsp-formatting tsserver`). By passing the server name to the command, there will never be a prompt.

krobelus replied:
ok a prompt sounds like a reasonable default UI.
As for how to force a specific set of servers, I'm not sure if arguments are the best way to do this.
Maybe we should have a `str-list` option that users can set before their request..

krobelus commented on src/language_features/hover.rs:
```diff
@@ -49,28 +77,43 @@ pub fn editor_hover(
                     if for_hover_buffer {
                         // We are typically creating Markdown, so use a standard Markdown enumerator.
-                        return format!("* {}", message);
+                        return format!("* [{server_name}] {message}");
```
this still shows "c_cpp" here which is not the server name I think.
Also if there is only one server (common case) we don't want to name it here

gbrlsnchs replied:
Yeah, I can add logic to omit such labels when there's a single server.

As for capabilities, are you okay with the output being ordered? I thought it'd be more consistent and easier to handle.

krobelus replied:
> As for capabilities, are you okay with the output being ordered? I thought it'd be more consistent and easier to handle.

you mean the `(serverA, serverB)` labels in lsp-capabilities output? Probably ideal to use the input order (or input ranking once we have that) but alphabetic order is okay too

gbrlsnchs replied:
I mean the capabilities per se. Previously, they were arbitrarily ordered (or maybe there was an implicit order), but because I had to use their names as a map key, I made my life easier by making the output consistent by using a BTreeMap.

If you wish the original order, I think we'd need to use something like [`indexmap`](https://github.com/bluss/indexmap).

krobelus replied:
it's fine to sort capabilities, the previous order seems pretty random

krobelus commented on src/context.rs:
```diff
@@ -33,76 +34,91 @@
     pub work_done_progress_report_timestamp: time::Instant,
-    pub pending_file_watchers: HashMap<Option<PathBuf>, Vec<CompiledFileSystemWatcher>>,
+    pub pending_file_watchers:
+        HashMap<(ServerName, String, Option<PathBuf>), Vec<CompiledFileSystemWatcher>>,
```
The "To server" and "From server" logs need to be adapted. Right now I can't tell which server the duplicate requests are directed at.

I noticed this when I tried

	[language.foo.c_cpp]
	filetypes = ["c", "cpp"]
	roots = ["compile_commands.json", ".clangd", ".git", ".hg"]
	command = "clangd"
	[language.foo.protobuf]
	filetypes = ["cpp"]
	roots = ["compile_commands.json", ".clangd", ".git", ".hg"]
	command = "pls"

trying to get go-to-definition for protobuf messages inside C++.
Sadly it doesn't work, clangd always responds before [pls](https://github.com/lasorda/protobuf-language-server/)

I think this one we could solve by ranking `pls` before `clangd`.
Defining `pls` as "primary" server feels odd but would work here.

gbrlsnchs replied:
Can kak-lsp currently handle gotos for multiple results? I know tsserver does return multiple results at some times. If so, we can go with merging the results instead of picking the first non-empty one. I think ranking them is not the ideal solution, because we would omit results anyway. What do you think?

gbrlsnchs replied:
Btw, I just remembered, the way I set up this was to use a `BTreeMap` for servers, therefore currently they are lexicographically ordered, but it should be simple to add a custom ordering logic that respects a ranking value.

As for gotos, it's possible to merge results, but on Kakoune side, it uses a single root. I think this root should come from the "priority" server.

krobelus replied:
> Can kak-lsp currently handle gotos for multiple results?

yeah in this case it will call `lsp-show-goto-choices` which shows a location list buffer.
We could maybe make that call lsp-menu instead

> If so, we can go with merging the results instead of picking the first non-empty one. I think ranking them is not the ideal solution, because we would omit results anyway. What do you think?

AFAICT the current behavior is the "first non-empty one" meaning it's a race between all language servers.
It would be good to make this deterministic. The simplest approach would be to wait for all responses (perhaps with a timeout) and then combine them. If that's too much work it can be a follow-up but let's keep that in mind.

> Btw, I just remembered, the way I set up this was to use a BTreeMap for servers, therefore currently they are lexicographically ordered, but it should be simple to add a custom ordering logic that respects a ranking value.

Good. We have two possible behaviors, either give the highest-ranked non-empty response, or merge all responses.
For my use case I'd like the first. Maybe there are uses for the second.

gbrlsnchs replied:
I pushed a commit that merges the results. For goto specifically, I think it's better UX to present all possible results. The logic for choosing the highest-ranked would be hard to mix with the merging strategy.

krobelus replied:
> For goto specifically, I think it's better UX to present all possible results.

yes let's show all of them for now, but remove duplicates

krobelus commented on src/session.rs:
```diff
@@ -130,10 +137,11 @@ pub fn start(config: &Config, initial_request: Option<String>) -> i32 {
                             controller_entry.insert(spawn_controller(
                                 config.clone(),
-                                route,
+                                language_id.clone(),
```
Haven't had time for a full review, some comments in passing:
I'm surprised we only spawn a controller for the first server.
Won't that lead to weird behavior if the configured servers have different root paths or language IDs?

gbrlsnchs replied:
It's not exactly for the first server, as the `controllers` key is now a vector of routes, and not a single route anymore. So in reality it spawns a controller for the first time a route sequence is seen by kak-lsp.

krobelus commented on src/session.rs:
```diff
@@ -211,9 +219,14 @@
-            info!("Exit {} in project {}", route.language, route.root);
+    controllers.retain(|routes, controller| {
+        if routes
+            .iter()
```
We should add an assertion somewhere that either all routes are from this session, or none is.
Because if only some are, then the behavior is weird.

krobelus commented on src/types.rs:
```diff
@@ -14,7 +14,8 @@ pub enum Void {}
 pub struct Config {
-    pub language: HashMap<String, LanguageConfig>,
+    #[serde(alias = "language_server")]
+    pub language: HashMap<ServerName, LanguageServerConfig>,
```
it would probably be better to rename this to `language_server` (also in the default kak-lsp.toml) and add the alias for the legacy name.
That can all go in the parent commit.

gbrlsnchs replied:
This has been addressed in the major commit overhaul I did.

krobelus commented on rc/lsp.kak:
```diff
@@ -1758,11 +1760,11 @@ define-command -hidden lsp-show-message-info -params 1 -docstring %{
 }
 
-define-command -hidden lsp-show-message-log -params 1 -docstring %{
+define-command -hidden lsp-show-message-log -params 2 -docstring %{
```
I'm not yet sure if this update is a good idea (because it might break users and complicate the interface for a niche use case).
If it is worth it, the other lsp-show-message-* should be updated in tandem.

gbrlsnchs replied:
When I was testing a setup with 3 different servers, this certainly helped with identifying which server was which in the `*debug*` buffer.

krobelus replied:
ok sorry I just searched on github; no one overrides these commands, so let's just update the other `lsp-show-message-*` commands as well

gbrlsnchs commented:
@krobelus I did my best to split commits in a meaningful way. After I split them, I rebased them so I could edit them in order to make ad-hoc changes that would make each commit buildable, which then I tested by running `git rebase -x 'cargo check' upstream/master`.

Here's how I feel after that:
![image](https://github.com/kak-lsp/kak-lsp/assets/7577128/4a847836-7029-47cc-a9df-055c832000f4)

There are some items in my to-do list in the PR description, which I'll now resume working on.

One major change I did was using filetypes as language IDs (instead of server name), and then I added a `language_ids` section at the config root that allows overriding language IDs if needed. That makes transition from old config even smoother, since there's no need to add a new field to make things work.

krobelus commented on src/types.rs:
```diff
@@ -23,12 +24,14 @@ pub struct Config {
     pub snippet_support: bool,
     #[serde(default)]
     pub semantic_tokens: SemanticTokenConfig,
+    #[serde(default)]
```
great idea! This is a nice simplification of our config.

krobelus commented on src/types.rs:
```diff
@@ -165,12 +169,13 @@
 pub struct Route {
     pub session: SessionId,
-    pub language: LanguageId,
+    pub server_name: ServerName,
```
(nit: the removal of `language` should happen in this commit (it's replaced by `server_name`))

krobelus commented on src/capabilities.rs:
```diff
@@ -527,80 +580,118 @@ pub fn server_has_capability(ctx: &Context, feature: &'static str) -> bool {
+                    &regopts.semantic_tokens_options.legend
+                }
+            };
+
```
Using the first server's offset encoding is wrong.
I noticed you correct this in a later commit.
But this has too much potential for confusion that I will continue reading the sum of all changes rather than the individual commits.
If you want to fix this structure (make sure that every commit passes all tests and doesn't have known flaws) that would help but else I can do this, no problem.
I appreciate the amount of work that has gone into this, it's a great exercise.

krobelus commented on README.asciidoc:
```diff
@@ -304,22 +304,52 @@ Please let us know if you have any ideas about how to make the default config mo
-same section is sent via `workspace/didChangeConfiguration`.  Additionally, kak-lsp will send
-arbitrary sections that are requested by the server in `workspace/configuration`.
+During server initialization, kak-lsp sends either the section specified by `settings_section`
+or the section with same name as the server; in this case `{"formatting.gofumpt":true}`.
```
> or the section with same name as the server

Hmmm, really? IIRC `settings_section` has no default value.
If I'm correct, then we need to add a `settings_section` to the `tailwindcss` example.
(Unfortunately the whole dynamic settings passing is super weird due to the lack of standardization in LSP.)

krobelus replied:
ok I think you're right, `settings_section` is optional

krobelus commented on README.asciidoc:
```diff
@@ -304,22 +304,52 @@
+----
+[language_ids]
+javascript = "javascriptreact"
+typescript = "typescriptreact"
```
I made another comment suggesting default language ID mappings.
Let's make sure this remains in sync with the default language ID mapping in kak-lsp.toml.

krobelus commented on README.asciidoc:
```diff
@@ -304,22 +304,52 @@
+
+[language_server.tailwindcss]
+filetypes = ["javascript", "typescript"]
+roots = ["tailwind.config.ts", "tailwind.config.js"]
```
are these config files required?

krobelus commented on README.asciidoc:
```diff
@@ -304,22 +304,52 @@
+filetypes = ["javascript", "typescript"]
+roots = ["tailwind.config.ts", "tailwind.config.js"]
+command = "tailwindcss-language-server"
+args = ["--stdio"]
```
nit: If other editors enable tailwindcss by default (no idea if they do) maybe we should too

krobelus commented on kak-lsp.toml:
```diff
@@ -6,163 +6,184 @@ verbosity = 2
+# This section overrides language IDs.
+# By default, kak-lsp use filetypes for the IDs.
+# See https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocumentItem
+[language_ids]
```
Hypothetically, if a user wanted to use clangd for the `c` and `c++` filetype but with different parameters each, that'd be possible now.
They can define `language_server.clangd-for-c` and `language_server.clangd-for-cpp`.
Nice.

krobelus commented on kak-lsp.toml:
```diff
@@ -6,163 +6,184 @@
 args = ["start"]
 
-[language.c_cpp]
+[language_server.clangd]
```
we used to send `c_cpp` always. Maybe we should keep doing that because I'm not sure if all *.h files are detected correctly.

krobelus commented on kak-lsp.toml:
```diff
@@ -6,163 +6,184 @@
-# # See https://github.dev/eclipse/eclipse.jdt.ls
-# # "java.format.insertSpaces" = true
-
-[language.jsx] # works for javascript as well
```
The reason why this uses `jsx` instead of `javascript` is a bit hidden,
because it was documented only for [`tsx` instead of `typescript`](https://github.com/kak-lsp/kak-lsp/commit/e8f109d7da52aa9a6c4586dcbc6d4bae2ae47c77).
Essentially `jsx` (perhaps that's the same as `javascriptreact`?) is a
superset of javascript, so it works for plain javascript.
On the other hand, `language_id=javascript` doesn't work for *.jsx files.
So I think what we should do here is add these mappings

	[language_ids]
	javascript = "jsx"
	typescript = "tsx"

unless there is a reason for changing the status quo.. I don't know much about frontend dev.
(I'm not sure if `javascriptreact` is better/worse)

krobelus replied:
ok looks like you're right; I've read elsewhere that `javascriptreact` is the proper language ID and not `jsx`

krobelus commented on kak-lsp.toml:
```diff
@@ -351,47 +326,76 @@
+filetypes = ["latex"]
+roots = [".git", ".hg"]
+command = "texlab"
+[language_server.texlab.settings.texlab]
```
I wonder if we should keep sorting by "primary filetype".
However I'm not sure, so no need to change this now.

krobelus commented on rc/lsp.kak:
```diff
@@ -1184,17 +1184,17 @@ $([ -z ${kak_hook_param+x} ] || echo hook = true)
-define-command lsp-formatting -docstring "Format document" %{
-    lsp-formatting-request false
+define-command lsp-formatting -params ..1 -docstring "Format document" %{
+    lsp-formatting-request false %arg{1}
```
Since this is public we should document the parameter in the docstring. Same below.
(Also in future we can equip this function with a `shell-script-candidates` completer that shows provides server names).

krobelus commented on rc/lsp.kak:
```diff
@@ -1226,17 +1227,17 @@
 
-define-command lsp-range-formatting -docstring "Format selections" %{
-    lsp-range-formatting-request false
+define-command lsp-range-formatting -params ..1 -docstring "Format selections" %{
```
let's use explicit `0..1` instead of `..1` for consistency with the rest of the code

krobelus commented on src/context.rs:
```diff
@@ -33,76 +42,83 @@ pub struct OutstandingRequests {
+        ),
+    >,
+    pub code_lenses: HashMap<String, Vec<(ServerName, CodeLens)>>,
+    pub completion_items: Vec<(ServerName, CompletionItem)>,
```
Have you given thought to moving these two into `ServerSettings`?
I guess that would be less convenient because of the way we forward it to the editor.

krobelus commented on src/context.rs:
```diff
@@ -111,58 +127,91 @@ impl Context {
             ops,
             Box::new(
-                move |ctx: &mut Context, meta: EditorMeta, mut results: Vec<R::Result>| {
-                    if let Some(result) = results.pop() {
```
The old results.pop() seems super weird. Why would we not consume all of results?
Maybe it's just an `Option<R::Result>` in disguise...
Anyway, the new version seems fine though it would be good to understand this.
Maybe it's because we have `batch_sizes()` now.

krobelus commented on src/controller.rs:
```diff
@@ -93,8 +121,44 @@ pub fn start(
+        let server_rxs: Vec<&Receiver<ServerMessage>> = language_servers
+            .values()
+            .map(|(transport, _)| transport.from_lang_server.receiver())
+            .collect();
```
I think `server_rxs` is loop-invariant, so we could initialize it before the loop?
I'd need to check.

krobelus commented on src/controller.rs:
```diff
@@ -104,65 +168,137 @@
                 // we park all requests from editor before initialization is complete
                 // and then dispatch them
-                if ctx.capabilities.is_some() {
+                let parked: Vec<_> = ctx
```
something for the future: if one server is parked and another one isn't, then we can already talk to some servers

krobelus commented on src/controller.rs:
```diff
@@ -171,69 +307,97 @@
-                                        },
+                                        }
                                         code => {
+                                            // TODO: Detail messages for different failing servers.
```
Is this TODO about the case where multiple servers return errors?

krobelus commented on src/controller.rs:
```diff
@@ -404,8 +568,12 @@
             goto::text_document_references(meta, params, ctx);
         }
         notification::Exit::METHOD => {
-            ctx.notify::<notification::Exit>(());
```
nit: maybe we should have a `ctx.notify_all` helper method.

krobelus commented on src/diagnostics.rs:
```diff
@@ -11,12 +11,24 @@ use lsp_types::*;
+        .map(|d| (server_name.clone(), d))
+        .collect();
+    diagnostics.extend(params);
+    ctx.diagnostics.insert(buffile.to_string(), diagnostics);
```
I guess this would be more elegant if we used `ctx.diagnostics.entry().or_insert()...` instead of remove() followed by insert(). Didn't try it though

krobelus commented on src/diagnostics.rs:
```diff
@@ -204,14 +221,16 @@
         write_response_to_fifo(meta, &ctx.diagnostics);
         return;
     }
+    let (_, main_settings) = ctx.language_servers.first_key_value().unwrap();
```
if we compute a "most general root path" then we don't need to fall back to the main settings.
That can be done later though

krobelus commented on src/diagnostics.rs:
```diff
@@ -257,13 +283,14 @@
                     format!(
-                        "{}:{}:{}: {}",
-                        short_file_path(filename, &ctx.root_path),
+                        "{}:{}:{}: ({server_name}) {}",
```
I think we should only add the server name if there are multiple servers configured

krobelus commented on src/language_features/call_hierarchy.rs:
```diff
@@ -32,56 +44,77 @@ fn request_call_hierarchy(
+    let result = results
+        .into_iter()
+        .find(|(_, response)| response.is_some())
+        .and_then(|(server_name, item)| Some((server_name, item.unwrap())));
```
I don't think we should call `unwrap()` here, at least the previous code didn't.
Of course in practice, most language servers will probably send an empty vector but I don't think the protocol guarantees that.
I'd write something like

	and_then(|...| item.into_iter().map(|item| (server_name, item)));

krobelus commented on src/language_features/call_hierarchy.rs:
```diff
@@ -32,56 +44,77 @@
-            params,
-            move |ctx: &mut Context, meta, result| {
-                format_call_hierarchy_calls(meta, ctx, incoming_or_outgoing, &item, &result);
+            RequestParams::Each(params),
```
Can we not use `RequestParams::All` here?

krobelus commented on src/language_features/call_hierarchy.rs:
```diff
@@ -32,56 +44,77 @@
-                format_call_hierarchy_calls(meta, ctx, incoming_or_outgoing, &item, &result);
+            RequestParams::Each(params),
+            move |ctx: &mut Context, meta, results| {
+                if let Some(result) = results.first() {
```
ok so this shows the results of whichever server responds first.
This will probably need correction (what if one server always returns `[]`?).
It can be done in a follow-up. Maybe a priority, or maybe skip empty reslts..

krobelus commented on src/language_features/code_action.rs:
```diff
@@ -33,85 +41,144 @@ pub fn text_document_code_action(meta: EditorMeta, params: EditorParams, ctx: &m
-        &document.text,
-        ctx.offset_encoding,
-    );
-    code_actions_for_range(meta, params, ctx, range)
```
this should be renamed to `code_actions_for_ranges`

krobelus commented on src/language_features/code_action.rs:
```diff
@@ -33,85 +41,144 @@
+        .map(|(server_name, _)| server_name)
+        .collect();
 
+    // TODO: Should pattern contain the server's name?
```
we can try that in future

krobelus commented on src/language_features/formatting.rs:
```diff
@@ -1,31 +1,77 @@
+                } else {
+                    "lsp-formatting"
+                };
+                let cmd = format!("{} {}", cmd, server_name);
```
should editor_quote `server_name`, just in case

krobelus commented on src/language_features/rename.rs:
```diff
@@ -11,26 +11,54 @@ use super::super::workspace;
+        meta,
+        RequestParams::Each(req_params),
+        move |ctx: &mut Context, meta, results| {
+            let result = match results.into_iter().find(|(_, v)| v.is_some()) {
```
ok so it's safe to broadcast the rename request because we'll apply only the first response

krobelus commented:
This is really fantastic, well designed and executed. Thanks!
I'll merge soon. If you want to address some review comments, go ahead, else I'll do the important ones.

There's a broken system test, here's the fix:

```diff
diff --git a/test/gopls-dynamic-settings.sh b/test/gopls-dynamic-settings.sh
index 454b55b..e0da1e3 100755
--- a/test/gopls-dynamic-settings.sh
+++ b/test/gopls-dynamic-settings.sh
@@ -29,7 +29,7 @@ test_tmux capture-pane -p
 
 echo '
 set global lsp_config %{
-	[language.go.settings.gopls]
+	[language_server.gopls.settings.gopls]
 	"formatting.gofumpt" = true
 }' | kak -p "$test_kak_session"
 test_sleep
```

krobelus commented on README.asciidoc:
```diff
@@ -304,22 +304,52 @@ Please let us know if you have any ideas about how to make the default config mo
 
 Many servers accept configuration options that are not part of the LSP spec.  The TOML table
-`[language.<filetype>.settings]` holds those configuration options.  It has the same structure
+`[language_server.<server_name>.settings]` holds those configuration options.  It has the same structure
```
I believe that if the old format (`[language.c_cpp]`) is used, then we should keep sending `c_cpp` as languageId, not the filetype. This is to avoid breaking existing configs. Unless we find that all filetypes work fine but I doubt that (for example with typescript-language-server, languageId javascript didn't work for jsx files last time I checked).
To implement this, we probabl need a custom deserializer.

krobelus commented:
I'm actually using this in production now, it feels awesome (I have a silly setup where I want to use two python language servers at the same time).
One other thing I noticed is that we should send codeLens requests only to `eligible_servers`.

raiguard commented:
Initial impressions are very good, thanks for working on this! Only problem I've run into is related to completions sorting. When using [FMTK](https://github.com/justarandomgeek/vscode-factoriomod-debug) and [LuaLS](https://github.com/LuaLS/lua-language-server/) together in a Lua file, the FMTK completions are buried under a mountain of plain text completions. 

LuaLS doesn't return any completions in this situation, so I assume that kak-lsp is erroneously letting the plaintext completions show up because of that. If I remove LuaLS, then the FMTK completions work properly and no plaintext completions are shown.

![image](https://github.com/kak-lsp/kak-lsp/assets/3515394/50ea7f75-e998-4a10-bd8a-6fb474d067a0)

gbrlsnchs commented:
Hey @krobelus, I'm a little busy lately and I haven't been able to give this PR some love. I just enabled maintainers to modify it as well, so please feel free to make adjustments as you seem fit!

gbrlsnchs commented:
> Initial impressions are very good, thanks for working on this! I hope we can add a combined diagnostics view to see all diagnostics for all servers in the future, but that can be a separate PR.

Hey @raiguard, thanks! I'm glad to help. Multiple diagnostics is already a thing. Each diagnostic is labeled by their respective language server. Is there a use case where they didn't properly show up to you?

> Turns out LuaLS returns text completions for strings, so this is theoretically working as intended. Something I would definitely do is sort text completions after all other completions, since more specific completions will undoubtedly be more useful.

This is something I also wish we had for kak-lsp. Maybe we can work on this as a future improvement, @krobelus? To let users choose their sorting priorities by item kind, maybe.

krobelus commented:
I'll tackle this next. Want to finish another thing first..

> Turns out LuaLS returns text completions for strings, so this is theoretically working as intended. Something I would definitely do is sort text completions after all other completions, since more specific completions will undoubtedly be more useful.

that first needs sorting support (as defined by LSP), which is added in https://github.com/kak-lsp/kak-lsp/pull/657 (needs a Kakoune patch as well)
Chances are it already works fine with that PR in most cases
If not, then it would be interesting to know how other editors solve it

krobelus commented:
@gbrlsnchs thanks for your patience, this is just awesome.
Merging now; let's keep our eyes open for things that don't work properly when using multiple language servers.
